onmt/modules/BiEncoder.lua

```diff
@@ -183,5 +186,11 @@ function BiEncoder:backward(batch, gradStatesOutput, gradContextOutput)
     gradContextBwd[{{}, t}]:copy(gradContextOutputBwd[{{}, batch.sourceLength - t + 1}])
   end
 
-  self.bwd:backward(batch, gradStatesOutputBwd, gradContextBwd)
+  local gradInputBwd = self.bwd:backward(batch, gradStatesOutputBwd, gradContextBwd)
+  -- gradInput
+  local gradInput = {}
+  for t = 1, batch.sourceLength do
+      gradInput[t] = torch.add(gradInputFwd[t], gradInputBwd[batch.sourceLength-t+1], 2)
```

Comment:
This is quite minor but can we avoid `torch.add` as it allocates new memory? You could add to `gradInputFwd` and return it.